### PR TITLE
Switch to Make.

### PR DESCRIPTION
**Pull request by bentley**
Packagers tend not to be fans of scons. [The Gentoo wiki lists some reasons](https://wiki.gentoo.org/wiki/SCons); at OpenBSD we share those concerns and have a few of our own.

Make is simple and widely portable. Here's a Makefile (compatible with at least BSD Make and GNU Make ≥ 4.0) that builds libgambatte and gambatte_sdl.

On my laptop, build time goes down from 53 seconds (scons) to 20 seconds (make).

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	libgambatte/src/video/sprite_mapper.o
+
+TEST_OBJECTS = \
+	test/testrunner.o
+
+all: $(SDL)
+
+SDL_LFLAGS != pkg-config --libs sdl
+$(SDL): $(SDL_OBJECTS) $(LIB)
+	$(CXX) $(CFLAGS) -o $@ -pthread $(SDL_OBJECTS) $(LIB) -lz \
```
Comment (hasufell) [inline Makefile]: This should be CXXFLAGS, also LDFLAGS are missing.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	libgambatte/src/video/next_m0_time.o \
+	libgambatte/src/video/ppu.o \
+	libgambatte/src/video/sprite_mapper.o
+
+TEST_OBJECTS = \
+	test/testrunner.o
+
+all: $(SDL)
+
+SDL_LFLAGS != pkg-config --libs sdl
```
Comment (hasufell) [inline Makefile]: I'd declare pkg-config in a variable PKG_CONFIG, that makes things easier for a lot of use cases (e.g. cross compiling with a wrapper script).

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+$(SDL): $(SDL_OBJECTS) $(LIB)
+	$(CXX) $(CFLAGS) -o $@ -pthread $(SDL_OBJECTS) $(LIB) -lz \
+		$(SDL_LFLAGS)
+
+$(LIB): $(LIB_OBJECTS)
+	$(AR) $(ARFLAGS) $@ $(LIB_OBJECTS)
+	ranlib $@
```
Comment (hasufell) [inline Makefile]: ranlib should also be declared in a variable

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+PKGCONFIG_CFLAGS != pkg-config --cflags sdl libpng
+.c.o:
+	$(CC) $(CFLAGS) -c $< -o $*.o
+.cpp.o:
+	$(CXX) -Ilibgambatte/src -Ilibgambatte/include -Igambatte_sdl/src \
+		-Icommon $(CFLAGS) -c $< -o $*.o \
```
Comment (hasufell) [inline Makefile]: should be CXXFLAGS

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,114 @@
+.SUFFIXES: .c .cpp .o
+
+LIB = libgambatte/libgambatte.a
+SDL = gambatte_sdl/gambatte_sdl
+TEST = test/testrunner
+
+PY = python
```
Comment (hasufell) [inline Makefile]: Some distros use the environment variable PYTHON to handle things. I also find the name better.
```
PYTHON ?= python
```
for example, but both ways work ofc.

**Comment by hasufell** [comment]
It seems both install and uninstall rules are missing. Would be good to have them too.

**Comment by bentley** [comment]
install is not a bad thing, but I didn’t bother with it because 1) there isn’t one already, and 2) it is harder to get right (handle DESTDIR? handle PREFIX? install library? install headers?…)

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+		$(SDL_LFLAGS) $(LDFLAGS)
+
+$(LIB): $(LIB_OBJECTS)
+	$(AR) $(ARFLAGS) $@ $(LIB_OBJECTS)
+	$(RANLIB) $@
+
+PKG_CONFIG = pkg-config
+PKGCONFIG_CFLAGS != $(PKG_CONFIG) --cflags sdl libpng
+.c.o:
+	$(CC) $(CXXFLAGS) -c $< -o $*.o
```
Comment (hasufell) [inline Makefile]: this should be $(CFLAGS) afais, because of CC (the other CXXFLAGS look correct though)

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+TEST_OBJECTS = \
+	test/testrunner.o
+
+all: $(SDL)
+
+SDL_LFLAGS != $(PKG_CONFIG) --libs sdl
+$(SDL): $(SDL_OBJECTS) $(LIB)
+	$(CXX) $(CXXFLAGS) -o $@ -pthread $(SDL_OBJECTS) $(LIB) -lz \
+		$(SDL_LFLAGS) $(LDFLAGS)
```
Comment (hasufell) [inline Makefile]: LDFLAGS (like ```-Wl,-O1```) only apply to libraries that follow them, so it should be earlier in the linker line... directly after CXXFLAG or after the '-o' target

File: Makefile [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+$(LIB): $(LIB_OBJECTS)
+	$(AR) $(ARFLAGS) $@ $(LIB_OBJECTS)
+	$(RANLIB) $@
+
+PKG_CONFIG = pkg-config
```
Comment (hasufell) [inline Makefile]: declared too late... it will make line 74 fail silently

**Comment by hasufell** [comment]
I'll prepare a PR to your repo

File: Makefile [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+	$(CXX) $(CXXFLAGS) $(LDFLAGS) -o $@ -pthread $(SDL_OBJECTS) $(LIB) $(ZLIB_LFLAGS) \
+		$(SDL_LFLAGS)
+
+$(LIB): $(LIB_OBJECTS)
+	$(AR) $(ARFLAGS) $@ $(LIB_OBJECTS)
+	$(RANLIB) $@
+
+PKGCONFIG_CFLAGS != $(PKG_CONFIG) --cflags sdl libpng zlib
+.c.o:
+	$(CC) $(CXXFLAGS) -c $< -o $*.o
```
Comment (hasufell) [inline Makefile]: oops, still missed this one :P

**Comment by hasufell** [comment]
any progress here?

**Comment by hasufell** [comment]
we could probably squash the commits, right @bentley

**Comment by bentley** [comment]
Squashed and rebased the commits.

**Comment by sergiobenrocha2** [comment]
Didn't work on debian based, sdl headers needs to be like this

#include <SDL/SDL.h>

In lot of files, and undefined references (missing -lSDL and others?)

**Comment by bentley** [comment]
Works for me on Jessie. Can you be more specific?

**Comment by sergiobenrocha2** [comment]
```
g++   -o gambatte_sdl/gambatte_sdl -pthread gambatte_sdl/src/audiosink.o gambatte_sdl/src/blitterwrapper.o gambatte_sdl/src/gambatte_sdl.o gambatte_sdl/src/parser.o gambatte_sdl/src/sdlblitter.o gambatte_sdl/src/str_to_sdlkey.o gambatte_sdl/src/usec.o common/adaptivesleep.o common/resample/src/chainresampler.o common/resample/src/i0.o common/resample/src/kaiser50sinc.o common/resample/src/kaiser70sinc.o common/resample/src/makesinckernel.o common/resample/src/resamplerinfo.o common/resample/src/u48div.o common/rateest.o common/skipsched.o common/videolink/rgb32conv.o common/videolink/vfilterinfo.o common/videolink/vfilters/catrom2x.o common/videolink/vfilters/catrom3x.o common/videolink/vfilters/kreed2xsai.o common/videolink/vfilters/maxsthq2x.o common/videolink/vfilters/maxsthq3x.o libgambatte/libgambatte.a \
		 
gambatte_sdl/src/audiosink.o: In function `(anonymous namespace)::openAudio(long, unsigned long, void (*)(void*, unsigned char*, int), void*)':
audiosink.cpp:(.text+0xb9): undefined reference to `SDL_OpenAudio'
audiosink.cpp:(.text+0xc5): undefined reference to `SDL_GetError'
gambatte_sdl/src/audiosink.o: In function `(anonymous namespace)::LockGuard::LockGuard(SDL_mutex*)':
audiosink.cpp:(.text+0x120): undefined reference to `SDL_mutexP'
gambatte_sdl/src/audiosink.o: In function `(anonymous namespace)::LockGuard::LockDeleter::del(SDL_mutex*)':
audiosink.cpp:(.text+0x15b): undefined reference to `SDL_mutexV'
gambatte_sdl/src/audiosink.o: In function `AudioSink::AudioSink(long, int, int)':
audiosink.cpp:(.text+0x1f1): undefined reference to `SDL_CreateMutex'
audiosink.cpp:(.text+0x20c): undefined reference to `SDL_CreateCond'
gambatte_sdl/src/audiosink.o: In function `AudioSink::~AudioSink()':
audiosink.cpp:(.text+0x2f7): undefined reference to `SDL_PauseAudio'
audiosink.cpp:(.text+0x2fc): undefined reference to `SDL_CloseAudio'
gambatte_sdl/src/audiosink.o: In function `AudioSink::write(short const*, unsigned long)':
audiosink.cpp:(.text+0x4f4): undefined reference to `SDL_CondWait'
gambatte_sdl/src/audiosink.o: In function `AudioSink::read(unsigned char*, unsigned long)':
audiosink.cpp:(.text+0x654): undefined reference to `SDL_CondSignal'
gambatte_sdl/src/audiosink.o: In function `AudioSink::SdlDeleter::del(SDL_mutex*)':
audiosink.cpp:(.text._ZN9AudioSink10SdlDeleter3delEP9SDL_mutex[_ZN9AudioSink10SdlDeleter3delEP9SDL_mutex]+0x14): undefined reference to `SDL_DestroyMutex'
gambatte_sdl/src/audiosink.o: In function `AudioSink::SdlDeleter::del(SDL_cond*)':
audiosink.cpp:(.text._ZN9AudioSink10SdlDeleter3delEP8SDL_cond[_ZN9AudioSink10SdlDeleter3delEP8SDL_cond]+0x14): undefined reference to `SDL_DestroyCond'
gambatte_sdl/src/gambatte_sdl.o: In function `(anonymous namespace)::SdlIniter::SdlIniter()':
gambatte_sdl.cpp:(.text+0x110a): undefined reference to `SDL_Init'
gambatte_sdl.cpp:(.text+0x1125): undefined reference to `SDL_GetError'
gambatte_sdl/src/gambatte_sdl.o: In function `(anonymous namespace)::SdlIniter::~SdlIniter()':
gambatte_sdl.cpp:(.text+0x1155): undefined reference to `SDL_Quit'
gambatte_sdl/src/gambatte_sdl.o: In function `(anonymous namespace)::JsOpen::~JsOpen()':
gambatte_sdl.cpp:(.text+0x1197): undefined reference to `SDL_JoystickClose'
gambatte_sdl/src/gambatte_sdl.o: In function `(anonymous namespace)::GambatteSdl::exec(int, char const* const*)':
gambatte_sdl.cpp:(.text+0x206d): undefined reference to `SDL_InitSubSystem'
gambatte_sdl.cpp:(.text+0x2086): undefined reference to `SDL_GetError'
gambatte_sdl.cpp:(.text+0x20ef): undefined reference to `SDL_JoystickEventState'
gambatte_sdl.cpp:(.text+0x2159): undefined reference to `SDL_ShowCursor'
gambatte_sdl.cpp:(.text+0x2168): undefined reference to `SDL_WM_SetCaption'
gambatte_sdl/src/gambatte_sdl.o: In function `(anonymous namespace)::GambatteSdl::handleEvents(BlitterWrapper&)':
gambatte_sdl.cpp:(.text+0x2abd): undefined reference to `SDL_PollEvent'
gambatte_sdl/src/gambatte_sdl.o: In function `(anonymous namespace)::GambatteSdl::run(long, int, int, ResamplerInfo const&, BlitterWrapper&)':
gambatte_sdl.cpp:(.text+0x2c12): undefined reference to `SDL_GetKeyState'
gambatte_sdl.cpp:(.text+0x2c35): undefined reference to `SDL_PauseAudio'
gambatte_sdl/src/gambatte_sdl.o: In function `(anonymous namespace)::JsOpen::JsOpen<std::_Rb_tree_const_iterator<unsigned char> >(std::_Rb_tree_const_iterator<unsigned char>, std::_Rb_tree_const_iterator<unsigned char>)':
gambatte_sdl.cpp:(.text+0x348a): undefined reference to `SDL_JoystickOpen'
gambatte_sdl/src/sdlblitter.o: In function `SdlBlitter::SdlBlitter(unsigned int, unsigned int, int, bool, bool)':
sdlblitter.cpp:(.text+0x35): undefined reference to `SDL_GetVideoInfo'
sdlblitter.cpp:(.text+0x68): undefined reference to `SDL_SetVideoMode'
sdlblitter.cpp:(.text+0xcb): undefined reference to `SDL_CreateRGBSurface'
sdlblitter.cpp:(.text+0x11b): undefined reference to `SDL_CreateYUVOverlay'
sdlblitter.cpp:(.text+0x161): undefined reference to `SDL_LockYUVOverlay'
gambatte_sdl/src/sdlblitter.o: In function `SdlBlitter::present()':
sdlblitter.cpp:(.text+0x41b): undefined reference to `SDL_UnlockYUVOverlay'
sdlblitter.cpp:(.text+0x43a): undefined reference to `SDL_DisplayYUVOverlay'
sdlblitter.cpp:(.text+0x452): undefined reference to `SDL_LockYUVOverlay'
sdlblitter.cpp:(.text+0x48a): undefined reference to `SDL_UpdateRect'
gambatte_sdl/src/sdlblitter.o: In function `SdlBlitter::toggleFullScreen()':
sdlblitter.cpp:(.text+0x4e1): undefined reference to `SDL_SetVideoMode'
gambatte_sdl/src/sdlblitter.o: In function `SdlBlitter::SurfaceDeleter::del(SDL_Surface*)':
sdlblitter.cpp:(.text._ZN10SdlBlitter14SurfaceDeleter3delEP11SDL_Surface[_ZN10SdlBlitter14SurfaceDeleter3delEP11SDL_Surface]+0x14): undefined reference to `SDL_FreeSurface'
gambatte_sdl/src/sdlblitter.o: In function `SdlBlitter::SurfaceDeleter::del(SDL_Overlay*)':
sdlblitter.cpp:(.text._ZN10SdlBlitter14SurfaceDeleter3delEP11SDL_Overlay[_ZN10SdlBlitter14SurfaceDeleter3delEP11SDL_Overlay]+0x1b): undefined reference to `SDL_UnlockYUVOverlay'
sdlblitter.cpp:(.text._ZN10SdlBlitter14SurfaceDeleter3delEP11SDL_Overlay[_ZN10SdlBlitter14SurfaceDeleter3delEP11SDL_Overlay]+0x27): undefined reference to `SDL_FreeYUVOverlay'
gambatte_sdl/src/usec.o: In function `getusecs()':
usec.cpp:(.text+0x5): undefined reference to `SDL_GetTicks'
gambatte_sdl/src/usec.o: In function `usecsleep(unsigned long)':
usec.cpp:(.text+0x45): undefined reference to `SDL_Delay'
libgambatte/libgambatte.a(file_zip.o): In function `(anonymous namespace)::GzFile::GzFile(char const*)':
file_zip.cpp:(.text+0x490): undefined reference to `gzopen'
file_zip.cpp:(.text+0x503): undefined reference to `gzread'
libgambatte/libgambatte.a(file_zip.o): In function `(anonymous namespace)::GzFile::rewind()':
file_zip.cpp:(.text+0x641): undefined reference to `gzrewind'
libgambatte/libgambatte.a(file_zip.o): In function `(anonymous namespace)::GzFile::read(char*, unsigned long)':
file_zip.cpp:(.text+0x6b4): undefined reference to `gzread'
libgambatte/libgambatte.a(file_zip.o): In function `(anonymous namespace)::GzFile::close()':
file_zip.cpp:(.text+0x717): undefined reference to `gzclose'
libgambatte/libgambatte.a(unzip.o): In function `unzOpenCurrentFile3':
unzip.c:(.text+0x1fb8): undefined reference to `inflateInit2_'
unzip.c:(.text+0x2065): undefined reference to `get_crc_table'
libgambatte/libgambatte.a(unzip.o): In function `unzReadCurrentFile':
unzip.c:(.text+0x265f): undefined reference to `crc32'
unzip.c:(.text+0x272d): undefined reference to `inflate'
unzip.c:(.text+0x2788): undefined reference to `crc32'
libgambatte/libgambatte.a(unzip.o): In function `unzCloseCurrentFile':
unzip.c:(.text+0x2aa4): undefined reference to `inflateEnd'
collect2: error: ld returned 1 exit status
make: *** [gambatte_sdl/gambatte_sdl] Error 1

```

**Comment by sergiobenrocha2** [comment]
*Ubuntu 14.04

**Comment by bentley** [comment]
Did you try installing pkg-config?

**Comment by sergiobenrocha2** [comment]
already installed

**Comment by bentley** [comment]
And what do `pkg-config --libs sdl` and `pkg-config --libs zlib` give? It looks very much like your pkg-config is not working properly.

**Comment by sergiobenrocha2** [comment]
pkg-config --libs sdl --> -lSDL
pkg-config --libs zlib --> -lz

**Comment by hasufell** [comment]
Try this patch

```
--- a/Makefile
+++ b/Makefile
@@ -87,8 +87,8 @@ TEST_OBJECTS = \

 all: $(SDL_TARGET)

-SDL_LFLAGS != $(PKG_CONFIG) --libs sdl
-ZLIB_LFLAGS != $(PKG_CONFIG) --libs zlib
+SDL_LFLAGS = $(shell $(PKG_CONFIG) --libs sdl)
+ZLIB_LFLAGS = $(shell $(PKG_CONFIG) --libs zlib)
 $(SDL_TARGET): $(SDL_OBJECTS) $(LIB)
        $(CXX) $(CXXFLAGS) $(LDFLAGS) -o $@ -pthread $(SDL_OBJECTS) $(LIB) \
                $(ZLIB_LFLAGS) $(SDL_LFLAGS)
@@ -97,7 +97,7 @@ $(LIB): $(LIB_OBJECTS)
        $(AR) $(ARFLAGS) $@ $(LIB_OBJECTS)
        $(RANLIB) $@

-PKGCONFIG_CFLAGS != $(PKG_CONFIG) --cflags sdl libpng zlib
+PKGCONFIG_CFLAGS = $(shell $(PKG_CONFIG) --cflags sdl libpng zlib)
 .c.o:
        $(CC) $(CFLAGS) -c $< -o $*.o
 .cpp.o:
@@ -120,7 +120,7 @@ test: $(TEST)
                test/hwtests/*/*/*/*.asm
        $(TEST) $(TEST_GBS)

-PNG_LFLAGS != $(PKG_CONFIG) --libs libpng
+PNG_LFLAGS = $(shell $(PKG_CONFIG) --libs libpng)
 $(TEST): $(TEST_OBJECTS) $(LIB)
        $(CXX) $(CXXFLAGS) -o $@ $(TEST_OBJECTS) $(LIB) \
                $(PNG_LFLAGS) $(ZLIB_LFLAGS)
```

**Comment by bentley** [comment]
Ah, that’s probably it. Unfortunately, neither `$(shell foo)` nor `!=` are portable. But `!=` is supported by GNU Make 4.0 and greater as well as BSD Make and there have been attempts to get it into POSIX, while `$(shell foo)` is only supported by GNU Make.

Once Ubuntu updates their GNU Make package, the problem will go away on its own.

**Comment by sergiobenrocha2** [comment]
"Once Ubuntu updates their GNU Make package, the problem will go away on its own."

Problem is ubuntu 14.04 is LTS... there are more 4 years of support.

**Comment by sergiobenrocha2** [comment]
The patch worked!

**Comment by sergiobenrocha2** [comment]
Nice, I was getting hardening-no-relro lintian error in sdl binary with scons, now with makefile it's fine.

Is it possible to merge this?

**Comment by hasufell** [comment]
> GNU Make 4.0 and greater as well as BSD Make and there have been attempts to get it into POSIX

just make BSD.Makefile and GNU.Makefile and call the correct one from within Makefile depending on some hacky checks... if these checks don't work, people can do `make -f BSD.Makefile` manually

**Comment by bentley** [comment]
> just make BSD.Makefile and GNU.Makefile and call the correct one from within Makefile depending on some hacky checks...

That’s exactly the opposite of my intention here… the point of switching to make is to be simple, fast, and portable. Duplicating makefiles and adding hackish checks over a single feature missing in a nearly two‐year‐old GNU Make is something I emphatically don’t want.

Here’s an alternative way without any hackish checks… GNU Make will read GNUmakefile by default, so we can set the variables in there with `$(shell foo)` and then include the regular Makefile. This will catch GNU Make 4.x as well, but that’s harmless. Other makes are unaffected.

**Comment by sergiobenrocha2** [comment]
Any news?

**Comment by bentley** [comment]
I just tweaked the compiler defaults bit. Other than that, I’ve been using this makefile unmodified for several months.

**Comment by andychase** [comment]
I like the simplicity of this but I'm not sure how well this would work for Windows builds (with the native VC compiler). Scons does a few conversions under the hood to make it work.

This might work with Mysys/Mingw but I don't know how to set that up.

**Comment by hasufell** [comment]
> I like the simplicity of this but I'm not sure how well this would work for Windows builds

This is an addition, not a replacement of SCons.

**Comment by andychase** [comment]
The title of this issue says "switch". Also it usually not a good idea to have duplicated logic unless there is a really good reason.

**Comment by hasufell** [comment]
> Also it usually not a good idea to have duplicated logic unless there is a really good reason.

I've been a distro package for more than 4 years and I can assure you that everyone hates SCons (especially source distros) since it needs various hacks to work correctly. And with "correctly" I don't just mean "it compiles". This Makefile makes things easier for that audience, so it's not really "duplicated logic".

Besides that, compilation time is reduced. Please see the initial post of this PR.

**Comment by HunterZ** [comment]
Banging my head on SCons right now because I am trying to cross-compile for an arm platform using a custom cross-compiler toolchain. SCons sucks at this use case.

Mark me down as another proponent of changing to a build system that doesn't suck. What's wrong with CMake? Gambatte is not doing anything fancy enough to make SCons is a huge advantage here...

**Comment by DMinsky** [comment]
I have had zero success in building this using _build_sdl.sh_ on Fedora 27 with following error:

`scons: Reading SConscript files ...
Checking for C library z... yes
TypeError: must be str, not bytes:
  File "/home/dminsky/Projects/gambatte/gambatte_sdl/SConstruct", line 51:
    version_str_def = [ 'GAMBATTE_SDL_VERSION_STR', r'\"r' + git_revno + r'\"' ]
`

But with _make_ it works like a charm.

**Comment by filpAM** [comment]
...Why is this still stuck on a limbo?

I agree having a Makefile would be really useful for gcc-toolchain users but I would rather not "Switch"(as the title suggest) but have it as an additional building tool, since sinamas and some other contribuitors still use scons.

**Comment by yoyz** [comment]
Really put a makefile next to the scons build things.
I really don't like to see scons, cmake stuff everywhere.
It's really cool when this tool work, but could be a nightmare when you try just to build the software and scons/cmake break.
And tell in the README that scons is the "preffered" way of building the software.
So you won't bother too much with support request.
People like me preffered to work with Makefile in the end.

**Comment by HunterZ** [comment]
Normally I'd prefer CMake because it's actively maintained and if done
right requires a minimum of fiddling, and is relatively easy to maintain.

Unfortunately CMake appears to require a lot of spoon-feeding when doing
cross compilation, which is probably a major use case for Gambatte.

GNU autotools supports cross compilation fairly well, but I strongly
dislike everything else about it.

I don't know much about scons.

**Pull request closed without merging**